### PR TITLE
Restructure experiment commit message logic

### PR DESCRIPTION
**Pull request by julieg18**
# `main` <= #3016 <= #3029 <= this

Resolves [review comments](https://github.com/iterative/vscode-dvc/pull/3016#pullrequestreview-1232853927) in #3016

### CODE REVIEW

File: extension/src/test/suite/extension.test.ts [hunk 1]
```diff
@@ -19,19 +19,14 @@ import { DvcReader } from '../../cli/dvc/reader'
 import expShowFixture from '../fixtures/expShow/base/output'
 import plotsDiffFixture from '../fixtures/plotsDiff/output'
 import * as Disposer from '../../util/disposable'
-import {
-  RegisteredCliCommands,
-  RegisteredCommands
-} from '../../commands/external'
+import { RegisteredCommands } from '../../commands/external'
```
Comment (julieg18) [inline extension/src/test/suite/extension.test.ts:22]: Saw these warnings when I ran `yarn lint`. I think they're leftover from #3035

File: extension/src/cli/git/reader.ts [hunk 2]
```diff
@@ -31,18 +31,28 @@ export class GitReader extends GitCli {
...
+    [sha: string]: string
+  }> {
     const options = getOptions(
       cwd,
       Command.LOG,
       Flag.PRETTY_FORMAT_COMMIT_MESSAGE,
       Flag.NUMBER,
-      '1',
-      sha
+      '3',
```
Comment (julieg18) [inline extension/src/cli/git/reader.ts:42]: Not sure why but `dvc exp show -n 3` doesn't actually show the last three commits for our current demo project, but the 1st, 2nd, and 4th commits 🤔 

image

Comment (mattseddon) [reply extension/src/cli/git/reader.ts:42]: looks like `exp show` is only showing merges and possibly the first parent:

```
git log --pretty="format:%H %B" HEAD~5..HEAD -z --merges --first-parent
```

This causes an issue where we cannot reliably grab the required commits by just going X commits back in time. We will need to pass the oldest commit into `git log` along with the required options. You should be able to get the commit by taking the last entry from `Object.keys` of `exp show` 😢.

File: extension/src/test/suite/extension.test.ts (hunk 1 above)
Comment (mattseddon) [reply extension/src/test/suite/extension.test.ts:22]: I followed up in #3039

**Comment by julieg18** [comment]
@mattseddon Can you look over the changes when you have a chance?

File: extension/src/cli/git/reader.ts [hunk 3]
```diff
@@ -31,18 +31,28 @@ export class GitReader extends GitCli {
...
+      Flag.SEPARATE_WITH_NULL
     )
     try {
-      return await this.executeProcess(options)
-    } catch {}
+      const output = await this.executeProcess(options)
```
Comment (mattseddon) [inline extension/src/cli/git/reader.ts:46]: I would keep this code as dumb as possible. Return a string and let `collectExperiments` deal with it. We want to try and keep all of the parsing logic in one place.

Comment (mattseddon) [reply extension/src/cli/git/reader.ts:46]: Parsing the string somewhere in `addCommitDataToBranches` would make sense to me

Comment (mattseddon) [reply extension/src/cli/git/reader.ts:46]: Would be good to add a test for the parsing logic after it gets moved too.

**Review by mattseddon — APPROVED** [review]
> @mattseddon Can you look over these changes when you have a chance?

Looks good. 

Good thing you checked the demo project as the current logic needs to be updated. Please take a look the comments that I've left and LMK if you need any clarification.

File: extension/src/cli/git/reader.ts (hunk 2 above)
Comment (julieg18) [reply extension/src/cli/git/reader.ts:42]: > You should be able to get the commit by taking the last entry from Object.keys of exp show 😢.

Good idea! I'll go ahead and increase the `-n` from 3 to 6 so there is less of a chance of it happening, merge, and fix it properly in a followup.

**Pull request merged**